libtenzir/src/series_builder.cpp

```diff
@@ -1161,70 +1161,92 @@ auto builder_ref::try_atom(detail::atom_view value) -> caf::expected<void> {
+                                             "accept value {}",
+                                             full_ty, value));
+        }
+        return value;
+      } else {
+        // TODO: We could consider allowing some conversions from enumeration.
+        // However, this code path is normally not taken anyway. For example,
+        // the JSON parser first has to resolve the enumeration string, which
+        // requires this having enumeration type.
+        return caf::make_error(ec::convert_error,
+                               fmt::format("cannot convert enumeration to {}",
+                                           type_kind::of<ToType>));
+      }
+    } else if constexpr (std::same_as<ToType, duration_type>) {
+      // TODO: Should we prefer to error if no unit was specified?
+      auto unit = full_ty.attribute("unit").value_or("s");
+      if constexpr (
+        // TODO: These special cases were extracted from `cast.hpp`.
+        // We should make it so that this is not necessary.
+        std::same_as<FromType, int64_type>
+        || std::same_as<FromType, uint64_type>
+        || std::same_as<FromType, double_type>) {
+        return cast_value(FromType{}, value, ty, unit);
+      } else if constexpr (std::same_as<FromType, string_type>) {
+        auto result = cast_value(FromType{}, value, ty);
+        if (not result) {
+          result
+            = cast_value(FromType{}, fmt::format("{} {}", value, unit), ty);
         }
-      } else if constexpr (std::same_as<ToType, duration_type>) {
-        // TODO: Should we prefer to error if no unit was specified?
-        auto unit = full_ty.attribute("unit").value_or("s");
-        if constexpr (
-          // TODO: These special cases were extracted from `cast.hpp`.
-          // We should make it so that this is not necessary.
-          std::same_as<FromType, int64_type>
-          || std::same_as<FromType, uint64_type>
-          || std::same_as<FromType, double_type>) {
-          result = cast_value(FromType{}, value, ty, unit);
-        } else if constexpr (std::same_as<FromType, string_type>) {
-          result = cast_value(FromType{}, value, ty);
-          if (not result) {
-            result
-              = cast_value(FromType{}, fmt::format("{} {}", value, unit), ty);
+        return result;
+      }
+    } else if constexpr (std::same_as<ToType, time_type>) {
+      if constexpr (std::same_as<FromType, int64_type>
+                    || std::same_as<FromType, uint64_type>
+                    || std::same_as<FromType, double_type>) {
+        auto unit = full_ty.attribute("since_epoch_unit");
+        if (unit) {
+          auto since_epoch
+            = cast_value(FromType{}, value, duration_type{}, *unit);
+          if (not since_epoch) {
+            return since_epoch.error();
           }
-        } else {
-          result = cast_value(FromType{}, value, ty);
+          return time{} + *since_epoch;
```

Comment:
This should be the only relevant change, the rest is refactoring.

Reply:
Thanks, that made the review a whole lot easier.